Source/WebCore/Modules/permissions/Permissions.cpp

```diff
@@ -77,66 +82,77 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
+    auto* document = dynamicDowncast<Document>(*context);
 
-    auto context = m_navigator ? m_navigator->scriptExecutionContext() : nullptr;
-
-    if (!context || !context->globalObject()) {
-        promise.reject(Exception { InvalidStateError, "The context is invalid"_s });
-        return;
-    }
-
-    auto* document = dynamicDowncast<Document>(context);
-    if (!document) {
-        promise.reject(Exception { NotSupportedError });
-        return;
-    }
-
-    if (!document->isFullyActive()) {
-        promise.reject(Exception { InvalidStateError, "The document is not fully active"_s });
-        return;
-    }
-
-    if (!document->page()) {
-        promise.reject(Exception { NotSupportedError });
-        return;
-    }
+    if (document) {
+        if (!document->isFullyActive()) {
+            promise.reject(Exception { InvalidStateError, "The document is not fully active"_s });
+            return;
+        }
 
-    if (!permissionDescriptorValue) {
-        promise.reject(Exception { DataError, "The parameter is invalid"_s });
-        return;
+        if (!document->page()) {
+            promise.reject(Exception { InvalidStateError, "The page does not exist"_s });
+            return;
+        }
     }
 
     JSC::VM& vm = context->globalObject()->vm();
     auto scope = DECLARE_THROW_SCOPE(vm);
-    auto parameterDescriptor = convert<IDLDictionary<PermissionDescriptor>>(*context->globalObject(), permissionDescriptorValue.get());
+    auto permissionDescriptor = convert<IDLDictionary<PermissionDescriptor>>(*context->globalObject(), permissionDescriptorValue.get());
     if (UNLIKELY(scope.exception())) {
         promise.reject(Exception { ExistingExceptionError });
         return;
     }
 
-    if (is<Document>(context) && !isAllowedByFeaturePolicy(downcast<Document>(*context), parameterDescriptor.name)) {
-        context->postTask([parameterDescriptor, promise = WTFMove(promise)](auto& context) mutable {
-            promise.resolve(PermissionStatus::create(context, PermissionState::Denied, parameterDescriptor));
-        });
+    if (document && !isAllowedByFeaturePolicy(*document, permissionDescriptor.name)) {
+        promise.resolve(PermissionStatus::create(*context, PermissionState::Denied, permissionDescriptor));
         return;
     }
 
     auto* origin = context->securityOrigin();
     auto originData = origin ? origin->data() : SecurityOriginData { };
 
-    PermissionController::shared().query(ClientOrigin { context->topOrigin().data(), originData }, PermissionDescriptor { parameterDescriptor }, *document->page(), [this, protectedThis = Ref { *this }, parameterDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
-        auto context = m_navigator ? m_navigator->scriptExecutionContext() : nullptr;
-        if (!context || !context->globalObject())
-            return;
+    if (auto* workerGlobalScope = dynamicDowncast<WorkerGlobalScope>(*context)) {
```

Comment:
I don't really understand the comment here. Rupin is first dealing with the worker case and early returning, and then dealing with the document case.

There is a little bit of document logic inside the worker case since we need to get to the document owning the worker to get its Page. However, the worker-owning document is not the same as `document` at the top-level of this function since it would be null in the worker case.

Reply:
There are 3 document checks outside workerGlobalScope:
```
if (document) {
         if (!document->isFullyActive()) {
```
and
`if (document && !isAllowedByFeaturePolicy(*document, permissionDescriptor.name))`
and
`ASSERT(document);`

Reply:
Oh, these checks. Very true. I agree we should move the document handling with those earlier document checks.

Reply:
Fixed!